### ISSUE TITLE
JS ABI Issue

### ISSUE THREAD

**Issue by joeykrug** [ethereum/web3.js#182]
Take the following contract for example:
```js
def createMarket(events:arr): return(events[0]
```

Its address is ```bf1f9a04cb5c85e9538a6f0e6ba15c824808cfee```
And its description is:
```js
[{ "name": "createMarket(int256[])", "type": "function", "inputs": [{ "name": "events", "type": "int256[]" }], "outputs": [{ "name": "out", "type": "int256" }] }]
```

If called in the web3.js console, e.g. 
```js
> contract.call({"from": eth.coinbase}).createMarket([-2333333333333333333333333333333333333333333333333332222222222233]).toNumber();```   it'll return ```-2.3333333333333333e+63
```
which is the correct value.

Now if we use this contract:
```js
def createMarket(branch, description:str, alpha, initialLiquidity, tradingFee, events:arr): return(events[0])
```
with address `0xe95aa1829447058201872db6ae3f04c3785a05b0`
And description:
```js
[{ "name": "createMarket(int256,string,int256,int256,int256,int256[])", "type": "function", "inputs": [{ "name": "branch", "type": "int256" }, { "name": "description", "type": "bytes" }, { "name": "alpha", "type": "int256" }, { "name": "initialLiquidity", "type": "int256" }, { "name": "tradingFee", "type": "int256" }, { "name": "events", "type": "int256[]" }], "outputs": [{ "name": "out", "type": "int256" }] }]
```

We end up getting completely wrong responses: 
```js
> contract.call({"from": eth.coinbase}).createMarket(2, "ok", 2, 2, 2, [444]).toNumber();
0
> contract.call({"from": eth.coinbase}).createMarket(2, "ok", 2, 2, 2, [-371]).toNumber();
65535
> contract.call({"from": eth.coinbase}).createMarket(2, "ok", 2, 2, 2, [371]).toNumber();
0
> contract.call({"from": eth.coinbase}).createMarket(2, "ok", 2, 2, 2, [-37117353667364131058659323780745282444029701754006041568106926382588839136039]).toNumber();
44528
> contract.call({"from": eth.coinbase}).createMarket(2, "ok", 2, 2, 2, [-4443434345]).toNumber();
65535
```

Edit: now contract.foo().call() is returning 
```js
error: could not decode, json: cannot unmarshal array into Go value of type struct { From string; To string; Value interface {}; Gas interface {}; GasPrice interface {}; Data string }
```

**Comment by frozeman** [comment]
i reformatted you issue it was unreadable. ;)

**Comment by frozeman** [comment]
web3.js contracts are solely for solidity at the moment and the contracts you're showing there are either incomplete or no solidity contracts i guess.

**Comment by joeykrug** [comment]
Oh yes it's a serpent contract --- 

Any reason it's giving this error?
```error: could not decode, json: cannot unmarshal array into Go value of type struct```

Even with a solidity contract that takes an array as a parameter it gives me that error when trying to call contract

**Comment by frozeman** [comment]
the problem is, i have without line numbers no way to tell where this error is generated.

**Comment by joeykrug** [comment]
Ok. Will run this again and get you some line numbers

**Comment by obscuren** [comment]
The calls generate very odd output. The `eth_call` or `eth_transact` look like `[[0xaaadress]]` which fails in decoding. It expects `[{"to":...,"from":..., etc}]`

**Comment by debris** [comment]
#### 1

> ```json
[{ "name": "createMarket(int256,string,int256,int256,int256,int256[])", "type": "function", "inputs": [{ "name": "branch", "type": "int256" }, { "name": "description", "type": "bytes" }, { "name": "alpha", "type": "int256" }, { "name": "initialLiquidity", "type": "int256" }, { "name": "tradingFee", "type": "int256" }, { "name": "events", "type": "int256[]" }], "outputs": [{ "name": "out", "type": "int256" }] }]
```

that's not correct. should be
 ```diff
- [{ "name": "createMarket(int256,string,int256,int256,int256,int256[])", "type": "function", "inputs": [{ "name": "branch", "type": "int256" }, { "name": "description", "type": "bytes" }, { "name": "alpha", "type": "int256" }, { "name": "initialLiquidity", "type": "int256" }, { "name": "tradingFee", "type": "int256" }, { "name": "events", "type": "int256[]" }], "outputs": [{ "name": "out", "type": "int256" }] }]
+ [{ "name": "createMarket(int256,bytes,int256,int256,int256,int256[])", "type": "function", "inputs": [{ "name": "branch", "type": "int256" }, { "name": "description", "type": "bytes" }, { "name": "alpha", "type": "int256" }, { "name": "initialLiquidity", "type": "int256" }, { "name": "tradingFee", "type": "int256" }, { "name": "events", "type": "int256[]" }], "outputs": [{ "name": "out", "type": "int256" }] }]
```

`name` of the function is used to generate function sha3. It cannot be `string` any more. Use bytes instead. Or even `bytes32`.

#### 2

parsing input should be ok. But just to be 100% sure, I've created [test](https://github.com/ethereum/web3.js/blob/develop/test/coder.encodeParam.js#L76-L86) for encoding you input params:
```javascript
test({ types: ['int', 'bytes', 'int', 'int', 'int', 'int[]'], values: [1, 'gavofyork', 2, 3, 4, [5, 6, 7]],
                                                            expected: '0000000000000000000000000000000000000000000000000000000000000009' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000003' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000001' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000002' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000003' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000004' +
                                                                      '6761766f66796f726b0000000000000000000000000000000000000000000000' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000005' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000006' +
                                                                      '0000000000000000000000000000000000000000000000000000000000000007'});
            
```

#### 3
> Latest output it gives:
```javascript
contract.createMarket.call([3])
```

You are right. I fixed it in latest web3.js develop. Here is [test](https://github.com/ethereum/web3.js/blob/develop/test/contract.js#L331-L364)

**Thanks for reporting it, cause it was critical.**

**Comment by chriseth** [comment]
Does this really work? The expectation should be
1 9 2 3 4 3 676176... 5 6 7
If this still causes problems, it would be another point in favour for moving to the new ABI proposal: https://github.com/ethereum/wiki/wiki/Proposal-for-new-ABI-value-encoding

**Comment by debris** [comment]
@chriseth why ```1 9 2 3 4 3 676176... 5 6 7``` ? 

I though length of dynamic params, should be always at the beginning. Look at [`sam` example](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI#examples)

**Comment by debris** [comment]
also, I though that constants should be before dynamic contents, #85

**Comment by debris** [comment]
btw. the proposal looks good, but it would be nice, to have 2-3 examples of encoding to clarify everything.

**Comment by chriseth** [comment]
The examples on the wiki page of the specification seem to contradict the specification itself.
I hope that the new proposal does not need examples for clarification, only for illustration :-)

**Comment by joeykrug** [comment]
Hmmm when i call a contract, getting a different error now
`contract.double({from: eth.coinbase}).call('1010101')`

`new BigNumber() not a number: [object Object]`

**Comment by frozeman** [comment]
yes ;)

**Comment by chriseth** [comment]
Actually, I looked at the sam example again at it does not contradict the specification because all of its parameters are dynamic.

**Comment by debris** [comment]
I fixed old encoding/decoding in #187. @chriseth can you verify that this tests for [encoding](https://github.com/debris/ethereum.js/blob/encoding/test/coder.encodeParam.js) and [decoding](https://github.com/debris/ethereum.js/blob/encoding/test/coder.decodeParam.js) are correct?

**Comment by debris** [comment]
fixed in #187

**Comment by ethers** [comment]
update: this is changing / being fixed in #194
